# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [1]:
#Import your libraries
import pandas as pd
import pymysql
from sqlalchemy import create_engine


# Challenge 1 - Import and Describe the Dataset

#### In this challenge we will use the `austin_weather` data. 

#### First, import it into a data frame called `austin`. You can find the data in Ironhack's database (db: weather, table: austin_weather). 

In [2]:
# Your code here

# cnx = sqlite3.connect('database.sqlite')
# driver   = 'mysql+pymysql:'
# user     = 'localhost'
# password = '13Lametripas'
# ip       = '127.0.0.1'
# database = 'weather'
# connection_string = f'{driver}//{user}:{password}@{ip}/{database}'
# engine = create_engine(connection_string)

# Read the sql file
# query = open('IronHackDB.sql', 'r')

# connection == the connection to your database, in your case prob_db
# austin = pd.read_sql_query(query.read(),connection)

# connection_string = 
# engine = create_engine(connection_string)
# pd.read_sql('SHOW TABLES;', 'IronHackDB.sql')
# austin = pd.read_sql('SELECT * FROM austin_weather', engine)
# austin_original = pd.read_sql('SELECT * FROM austin_weather', engine)



austin = pd.read_csv('austin_weather.csv')
austin_original = pd.read_csv('austin_weather.csv')

In [3]:
austin.head()

Unnamed: 0        Date  TempHighF  TempAvgF  TempLowF DewPointHighF  \
0           0  2013-12-21         74        60        45            67   
1           1  2013-12-22         56        48        39            43   
2           2  2013-12-23         58        45        32            31   
3           3  2013-12-24         61        46        31            36   
4           4  2013-12-25         58        50        41            44   

  DewPointAvgF DewPointLowF HumidityHighPercent HumidityAvgPercent  ...  \
0           49           43                  93                 75  ...   
1           36           28                  93                 68  ...   
2           27           23                  76                 52  ...   
3           28           21                  89                 56  ...   
4           40           36                  86                 71  ...   

  SeaLevelPressureAvgInches SeaLevelPressureLowInches VisibilityHighMiles  \
0                     29.68                     29.59                  10   
1                     30.13                     29.87                  10   
2                     30.49                     30.41                  10   
3                     30.45                      30.3                  10   
4                     30.33                     30.27                  10   

  VisibilityAvgMiles VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                  7                  2          20          4          31   
1                 10                  5          16          6          25   
2                 10                 10           8          3          12   
3                 10                  7          12          4          20   
4                 10                  7          10          2          16   

  PrecipitationSumInches               Events  
0                   0.46  Rain , Thunderstorm  
1                      0                       
2                      0                       
3                      0                       
4                      T                       

[5 rows x 22 columns]

#### Next, describe the dataset you have loaded: 
- Look at the variables and their types
- Examine the descriptive statistics of the numeric variables 
- Look at the first five rows of all variables to evaluate the categorical variables as well

In [4]:
# Your code here
austin.describe()

Unnamed: 0    TempHighF     TempAvgF     TempLowF
count  1319.000000  1319.000000  1319.000000  1319.000000
mean    659.000000    80.862775    70.642911    59.902957
std     380.906813    14.766523    14.045904    14.190648
min       0.000000    32.000000    29.000000    19.000000
25%     329.500000    72.000000    62.000000    49.000000
50%     659.000000    83.000000    73.000000    63.000000
75%     988.500000    92.000000    83.000000    73.000000
max    1318.000000   107.000000    93.000000    81.000000

In [14]:
# Your code here
len(austin.columns)

22

In [15]:
# Your code here
austin.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 22 columns):
Unnamed: 0                    1319 non-null int64
Date                          1319 non-null object
TempHighF                     1319 non-null int64
TempAvgF                      1319 non-null int64
TempLowF                      1319 non-null int64
DewPointHighF                 1319 non-null object
DewPointAvgF                  1319 non-null object
DewPointLowF                  1319 non-null object
HumidityHighPercent           1319 non-null object
HumidityAvgPercent            1319 non-null object
HumidityLowPercent            1319 non-null object
SeaLevelPressureHighInches    1319 non-null object
SeaLevelPressureAvgInches     1319 non-null object
SeaLevelPressureLowInches     1319 non-null object
VisibilityHighMiles           1319 non-null object
VisibilityAvgMiles            1319 non-null object
VisibilityLowMiles            1319 non-null object
WindHighMPH                   

#### Given the information you have learned from examining the dataset, write down three insights about the data in a markdown cell below

#### Your Insights:

1. There are 21 variables in the dataset. 3 of them are numeric and the rest contain some text.

2. The average temperature in Austin ranged between around 70 degrees F and around 93 degrees F. The highest temperature observed during this period was 107 degrees F and the lowest was 19 degrees F.

3. When we look at the head function, we see that a lot of variables contain numeric data even though these columns are of object type. This means we might have to do some data cleansing.


#### Let's examine the DewPointAvgF variable by using the `unique()` function to list all unique values in this dataframe.

Describe what you find in a markdown cell below the code. What did you notice? What do you think made Pandas to treat this column as *object* instead of *int64*? 

In [33]:
# Your code here
austin.DewPointAvgF.unique()

array(['49', '36', '27', '28', '40', '39', '41', '26', '42', '22', '48',
       '32', '8', '11', '45', '55', '61', '37', '47', '25', '23', '20',
       '33', '30', '29', '17', '14', '13', '54', '59', '15', '24', '34',
       '35', '57', '50', '53', '60', '46', '56', '51', '31', '38', '62',
       '43', '63', '64', '67', '66', '58', '70', '68', '65', '69', '71',
       '72', '-', '73', '74', '21', '44', '52', '12', '75', '76', '18'],
      dtype=object)

In [ ]:
# Your observation here
'''
One value is '-', which means it's nan. That's why Pandas is treating it as an object instead of int64.

'''

The following is a list of columns misrepresented as `object`. Use this list to convert the columns to numeric using the `pandas.to_numeric` function in the next cell. If you encounter errors in converting strings to numeric values, you need to catch those errors and force the conversion by supplying `errors='coerce'` as an argument for `pandas.to_numeric`. Coercing will replace non-convertable elements with `NaN` which represents an undefined numeric value. This makes it possible for us to conveniently handle missing values in subsequent data processing.

*Hint: you may use a loop to change one column at a time but it is more efficient to use `apply`.*

In [35]:
wrong_type_columns = ['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches']

In [47]:
# Your code here
austin[wrong_type_columns] = austin[wrong_type_columns].apply(pd.to_numeric, errors='coerce')

#### Check if your code has worked by printing the data types again. You should see only two `object` columns (`Date` and `Events`) now. All other columns should be `int64` or `float64`.

In [48]:
# Your code here
austin.dtypes

Unnamed: 0                      int64
Date                           object
TempHighF                       int64
TempAvgF                        int64
TempLowF                        int64
DewPointHighF                 float64
DewPointAvgF                  float64
DewPointLowF                  float64
HumidityHighPercent           float64
HumidityAvgPercent            float64
HumidityLowPercent            float64
SeaLevelPressureHighInches    float64
SeaLevelPressureAvgInches     float64
SeaLevelPressureLowInches     float64
VisibilityHighMiles           float64
VisibilityAvgMiles            float64
VisibilityLowMiles            float64
WindHighMPH                   float64
WindAvgMPH                    float64
WindGustMPH                   float64
PrecipitationSumInches        float64
Events                         object
dtype: object

# Challenge 2 - Handle the Missing Data

#### Now that we have fixed the type mismatch, let's address the missing data.

By coercing the columns to numeric, we have created `NaN` for each cell containing characters. We should choose a strategy to address these missing data.

The first step is to examine how many rows contain missing data.

We check how much missing data we have by applying the `.isnull()` function to our dataset. To find the rows with missing data in any of its cells, we apply `.any(axis=1)` to the function. `austin.isnull().any(axis=1)` will return a column containing true if the row contains at least one missing value and false otherwise. Therefore we must subset our dataframe with this column. This will give us all rows with at least one missing value. 

#### In the next cell, identify all rows containing at least one missing value. Assign the dataframes with missing values to a variable called `missing_values`.

In [55]:
# Your code here
missing_values = austin[austin.isnull().any(axis=1)]

There are multiple strategies to handle missing data. Below lists the most common ones data scientists use:

* Removing all rows or all columns containing missing data. This is the simplest strategy. It may work in some cases but not others.

* Filling all missing values with a placeholder value. 
    * For categorical data, `0`, `-1`, and `9999` are some commonly used placeholder values. 
    * For continuous data, some may opt to fill all missing data with the mean. This strategy is not optimal since it can increase the fit of the model.

* Filling the values using some algorithm. 

#### In our case, we will use a hybrid approach which is to first remove the data that contain most missing values then fill in the rest of the missing values with the *linear interpolation* algorithm.

#### Next, count the number of rows of `austin` and `missing_values`.

In [56]:
# Your code here
len(missing_values)

136

In [53]:
len(austin)

1319

#### Calculate the ratio of missing rows to total rows

In [57]:
# Your code here
136/1319

0.10310841546626232

As you can see, there is a large proportion of missing data (over 10%). Perhaps we should evaluate which columns have the most missing data and remove those columns. For the remaining columns, we will perform a linear approximation of the missing data.

We can find the number of missing rows in each column using the `.isna()` function. We then chain the `.sum` function to the `.isna()` function and find the number of missing rows per column

In [59]:
# Your code here
austin.isna().sum()

Unnamed: 0                      0
Date                            0
TempHighF                       0
TempAvgF                        0
TempLowF                        0
DewPointHighF                   7
DewPointAvgF                    7
DewPointLowF                    7
HumidityHighPercent             2
HumidityAvgPercent              2
HumidityLowPercent              2
SeaLevelPressureHighInches      3
SeaLevelPressureAvgInches       3
SeaLevelPressureLowInches       3
VisibilityHighMiles            12
VisibilityAvgMiles             12
VisibilityLowMiles             12
WindHighMPH                     2
WindAvgMPH                      2
WindGustMPH                     4
PrecipitationSumInches        124
Events                          0
dtype: int64

#### As you can see from the output, the majority of missing data is in one column called `PrecipitationSumInches`. What's the number of missing values in this column in ratio to its total number of rows?

In [ ]:
# Your code here
124

Over 10% data missing! Therefore, we prefer to remove this column instead of filling its missing values. It doesn't make sense to *guess* its missing values because the estimation will be too 

#### Remove this column from `austin` using the `.drop()` function. Use the `inplace=True` argument.

*Hints:*

* By supplying `inplace=True` to `drop()`, the original dataframe object will be changed in place and the function will return `None`. In contrast, if you don't supply `inplace=True`, which is equivalent to supplying `inplace=False` because `False` is the default value, the original dataframe object will be kept and the function returns a copy of the transformed dataframe object. In the latter case, you'll have to assign the returned object back to your variable.

* Also, since you are dropping a column instead of a row, you'll need to supply `axis=1` to `drop()`.

[Reference for `pandas.DataFrame.drop`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

In [63]:
# Your code here 
austin.drop(columns = 'PrecipitationSumInches', inplace=True, axis=1)

# Print `austin` to confirm the column is indeed removed

austin.head()

Unnamed: 0        Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  \
0           0  2013-12-21         74        60        45           67.0   
1           1  2013-12-22         56        48        39           43.0   
2           2  2013-12-23         58        45        32           31.0   
3           3  2013-12-24         61        46        31           36.0   
4           4  2013-12-25         58        50        41           44.0   

   DewPointAvgF  DewPointLowF  HumidityHighPercent  HumidityAvgPercent  ...  \
0          49.0          43.0                 93.0                75.0  ...   
1          36.0          28.0                 93.0                68.0  ...   
2          27.0          23.0                 76.0                52.0  ...   
3          28.0          21.0                 89.0                56.0  ...   
4          40.0          36.0                 86.0                71.0  ...   

   SeaLevelPressureHighInches  SeaLevelPressureAvgInches  \
0                       29.86                      29.68   
1                       30.41                      30.13   
2                       30.56                      30.49   
3                       30.56                      30.45   
4                       30.41                      30.33   

   SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
0                      29.59                 10.0                 7.0   
1                      29.87                 10.0                10.0   
2                      30.41                 10.0                10.0   
3                      30.30                 10.0                10.0   
4                      30.27                 10.0                10.0   

   VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  \
0                 2.0         20.0         4.0         31.0   
1                 5.0         16.0         6.0         25.0   
2                10.0          8.0         3.0         12.0   
3                 7.0         12.0         4.0         20.0   
4                 7.0         10.0         2.0         16.0   

                Events  
0  Rain , Thunderstorm  
1                       
2                       
3                       
4                       

[5 rows x 21 columns]

#### Next we will perform linear interpolation of the missing data.

This means that we will use a linear algorithm to estimate the missing data. Linear interpolation assumes that there is a straight line between the points and the missing point will fall on that line. This is a good enough approximation for weather related data. Weather related data is typically a time series. Therefore, we do not want to drop rows from our data if possible. It is prefereable to estimate the missing values rather than remove the rows. However, if you have data from a single point in time, perhaps a better solution would be to remove the rows. 

If you would like to read more about linear interpolation, you can do so [here](https://en.wikipedia.org/wiki/Linear_interpolation).

In the following cell, use the `.interpolate()` function on the entire dataframe. This time pass the `inplace=False` argument to the function and assign the interpolated dataframe to a new variable called `austin_fixed` so that we can compare with `austin`.

In [64]:
# Your code here
austin_fixed = austin.interpolate(inplace=False)

#### Check to make sure `austin_fixed` contains no missing data. Also check `austin` - it still contains missing data.

In [65]:
# Your code here
austin_fixed.isnull().any()

Unnamed: 0                    False
Date                          False
TempHighF                     False
TempAvgF                      False
TempLowF                      False
DewPointHighF                 False
DewPointAvgF                  False
DewPointLowF                  False
HumidityHighPercent           False
HumidityAvgPercent            False
HumidityLowPercent            False
SeaLevelPressureHighInches    False
SeaLevelPressureAvgInches     False
SeaLevelPressureLowInches     False
VisibilityHighMiles           False
VisibilityAvgMiles            False
VisibilityLowMiles            False
WindHighMPH                   False
WindAvgMPH                    False
WindGustMPH                   False
Events                        False
dtype: bool

In [66]:
austin.isnull().any()

Unnamed: 0                    False
Date                          False
TempHighF                     False
TempAvgF                      False
TempLowF                      False
DewPointHighF                  True
DewPointAvgF                   True
DewPointLowF                   True
HumidityHighPercent            True
HumidityAvgPercent             True
HumidityLowPercent             True
SeaLevelPressureHighInches     True
SeaLevelPressureAvgInches      True
SeaLevelPressureLowInches      True
VisibilityHighMiles            True
VisibilityAvgMiles             True
VisibilityLowMiles             True
WindHighMPH                    True
WindAvgMPH                     True
WindGustMPH                    True
Events                        False
dtype: bool

# Challenge 3 - Processing the `Events` Column

#### Our dataframe contains one true text column - the Events column. We should evaluate this column to determine how to process it.

Use the `value_counts()` function to evaluate the contents of this column

In [71]:
# Your code here:
austin_fixed.Events.value_counts()

                             903
Rain                         192
Rain , Thunderstorm          137
Fog , Rain , Thunderstorm     33
Fog                           21
Thunderstorm                  17
Fog , Rain                    14
Rain , Snow                    1
Fog , Thunderstorm             1
Name: Events, dtype: int64

In [72]:
austin_fixed[austin_fixed.Events == 'Rain'].Events.head()

18    Rain
20    Rain
22    Rain
37    Rain
38    Rain
Name: Events, dtype: object

Reading the values of `Events` and reflecting what those values mean in the context of data, you realize this column indicates what weather events had happened in a particular day.

#### What is the largest number of events happened in a single day? Enter your answer in the next cell.

In [ ]:
# Your answer:
'''
The largest number of events is 3: Fog, Rain, Thunderstorm.
'''

#### We want to transform the string-type `Events` values to the numbers. This will allow us to apply machine learning algorithms easily.

How? We will create a new column for each type of events (i.e. *Rain*, *Snow*, *Fog*, *Thunderstorm*. In each column, we use `1` to indicate if the corresponding event happened in that day and use `0` otherwise.

Below we provide you a list of all event types. Loop the list and create a dummy column with `0` values for each event in `austin_fixed`. To create a new dummy column with `0` values, simply use `austin_fixed[event] = 0`.

In [76]:
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']

# Your code here

for i in event_list:
    austin_fixed[i] = 0

# Print your new dataframe to check whether new columns have been created:

austin.head()

Snow
Fog
Rain
Thunderstorm


Unnamed: 0        Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  \
0           0  2013-12-21         74        60        45           67.0   
1           1  2013-12-22         56        48        39           43.0   
2           2  2013-12-23         58        45        32           31.0   
3           3  2013-12-24         61        46        31           36.0   
4           4  2013-12-25         58        50        41           44.0   

   DewPointAvgF  DewPointLowF  HumidityHighPercent  HumidityAvgPercent  ...  \
0          49.0          43.0                 93.0                75.0  ...   
1          36.0          28.0                 93.0                68.0  ...   
2          27.0          23.0                 76.0                52.0  ...   
3          28.0          21.0                 89.0                56.0  ...   
4          40.0          36.0                 86.0                71.0  ...   

   VisibilityAvgMiles  VisibilityLowMiles  WindHighMPH  WindAvgMPH  \
0                 7.0                 2.0         20.0         4.0   
1                10.0                 5.0         16.0         6.0   
2                10.0                10.0          8.0         3.0   
3                10.0                 7.0         12.0         4.0   
4                10.0                 7.0         10.0         2.0   

   WindGustMPH               Events  Snow  Fog  Rain  Thunderstorm  
0         31.0  Rain , Thunderstorm     0    0     0             0  
1         25.0                          0    0     0             0  
2         12.0                          0    0     0             0  
3         20.0                          0    0     0             0  
4         16.0                          0    0     0             0  

[5 rows x 25 columns]

#### Next, populate the actual values in the dummy columns of  `austin_fixed`.

You will check the *Events* column. If its string value contains `Rain`, then the *Rain* column should be `1`. The same for `Snow`, `Fog`, and `Thunderstorm`.

*Hints:*

* Use [`pandas.Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.contains.html) to create the value series of each new column.

* What if the values you populated are booleans instead of numbers? You can cast the boolean values to numbers by using `.astype(int)`. For instance, `pd.Series([True, True, False]).astype(int)` will return a new series with values of `[1, 1, 0]`.

In [99]:
# Your code here
for i in event_list:
    austin_fixed[[i]] = austin_fixed.Events.str.contains(i, regex=True).astype(int)


#### Print out `austin_fixed` to check if the event columns are populated with the intended values

In [104]:
# Your code here
austin_fixed[austin_fixed.Fog == 1]

Unnamed: 0        Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  \
19            19  2014-01-09         70        62        53           60.0   
50            50  2014-02-09         76        59        41           53.0   
57            57  2014-02-16         76        68        59           61.0   
70            70  2014-03-01         82        68        53           65.0   
84            84  2014-03-15         74        68        62           64.0   
...          ...         ...        ...       ...       ...            ...   
1119        1119  2017-01-13         78        67        56           69.0   
1123        1123  2017-01-17         59        54        48           55.0   
1127        1127  2017-01-21         81        65        48           58.0   
1142        1142  2017-02-05         72        62        51           64.0   
1262        1262  2017-06-05         90        78        66           70.0   

      DewPointAvgF  DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
19            55.0          50.0                 93.0                80.0   
50            45.0          38.0                 96.0                70.0   
57            57.0          53.0                 90.0                74.0   
70            57.0          48.0                 93.0                74.0   
84            62.0          59.0                 93.0                81.0   
...            ...           ...                  ...                 ...   
1119          63.0          51.0                100.0                74.0   
1123          48.0          46.0                100.0                86.0   
1127          51.0          33.0                100.0                62.0   
1142          59.0          51.0                100.0                89.0   
1262          68.0          65.0                100.0                72.0   

      ...  VisibilityAvgMiles  VisibilityLowMiles  WindHighMPH  WindAvgMPH  \
19    ...                 4.0                 0.0         12.0         3.0   
50    ...                 5.0                 0.0         13.0         4.0   
57    ...                 5.0                 0.0         13.0         5.0   
70    ...                 7.0                 0.0         16.0         5.0   
84    ...                 5.0                 0.0         12.0         4.0   
...   ...                 ...                 ...          ...         ...   
1119  ...                 8.0                 0.0         14.0         7.0   
1123  ...                 7.0                 0.0         14.0         6.0   
1127  ...                 7.0                 0.0         21.0         6.0   
1142  ...                 6.0                 0.0         12.0         3.0   
1262  ...                 9.0                 0.0         25.0         3.0   

      WindGustMPH                     Events  Snow  Fog  Rain  Thunderstorm  
19           19.0                        Fog     0    1     0             0  
50           20.0                        Fog     0    1     0             0  
57           23.0                        Fog     0    1     0             0  
70           27.0                        Fog     0    1     0             0  
84           17.0                 Fog , Rain     0    1     1             0  
...           ...                        ...   ...  ...   ...           ...  
1119         24.0                 Fog , Rain     0    1     1             0  
1123         22.0  Fog , Rain , Thunderstorm     0    1     1             1  
1127         38.0                        Fog     0    1     0             0  
1142         19.0                        Fog     0    1     0             0  
1262         41.0  Fog , Rain , Thunderstorm     0    1     1             1  

[69 rows x 25 columns]

#### If your code worked correctly, now we can drop the `Events` column as we don't need it any more.

In [105]:
# Your code here
austin_fixed.drop(columns = 'Events', inplace=True)

# Challenge 4 - Processing The `Date` Column

The `Date` column is another non-numeric field in our dataset. A value in that field looks like `'2014-01-06'` which consists of the year, month, and day connected with hyphens. One way to convert the date string to numerical is using a similar approach as we used for `Events`, namely splitting the column into numerical `Year`, `Month`, and `Day` columns. In this challenge we'll show you another way which is to use the Python `datetime` library's `toordinal()` function. Depending on what actual machine learning analysis you will conduct, each approach has its pros and cons. Our goal today is to practice data preparation so we'll skip the discussion here.

Here you can find the [reference](https://docs.python.org/3/library/datetime.html) and [example](https://stackoverflow.com/questions/39846918/convert-date-to-ordinal-python) for `toordinal`. The basic process is to first convert the string to a `datetime` object using `datetime.datetime.strptime`, then convert the `datetime` object to numerical using `toordinal`.

#### In the cell below, convert the `Date` column values from string to numeric values using `toordinal()`.

In [108]:
from datetime import datetime as dt

In [120]:
# Your code here
# d = dt.strptime('2010-03-01', '%Y-%m-%d').date()
# d.toordinal()

austin_fixed.Date = austin_fixed.Date.apply(lambda x: dt.strptime(x,"%Y-%m-%d").date().toordinal())

#### Print `austin_fixed` to check your `Date` column.

In [122]:
austin_fixed.head()

Unnamed: 0    Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  \
0           0  735223         74        60        45           67.0   
1           1  735224         56        48        39           43.0   
2           2  735225         58        45        32           31.0   
3           3  735226         61        46        31           36.0   
4           4  735227         58        50        41           44.0   

   DewPointAvgF  DewPointLowF  HumidityHighPercent  HumidityAvgPercent  ...  \
0          49.0          43.0                 93.0                75.0  ...   
1          36.0          28.0                 93.0                68.0  ...   
2          27.0          23.0                 76.0                52.0  ...   
3          28.0          21.0                 89.0                56.0  ...   
4          40.0          36.0                 86.0                71.0  ...   

   VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  WindHighMPH  \
0                 10.0                 7.0                 2.0         20.0   
1                 10.0                10.0                 5.0         16.0   
2                 10.0                10.0                10.0          8.0   
3                 10.0                10.0                 7.0         12.0   
4                 10.0                10.0                 7.0         10.0   

   WindAvgMPH  WindGustMPH  Snow  Fog  Rain  Thunderstorm  
0         4.0         31.0     0    0     1             1  
1         6.0         25.0     0    0     0             0  
2         3.0         12.0     0    0     0             0  
3         4.0         20.0     0    0     0             0  
4         2.0         16.0     0    0     0             0  

[5 rows x 24 columns]

# Challenge 5 - Sampling and Holdout Sets

#### Now that we have processed the data for machine learning, we will separate the data to test and training sets.

We first train the model using only the training set. We check our metrics on the training set. We then apply the model to the test set and check our metrics on the test set as well. If the metrics are significantly more optimal on the training set, then we know we have overfit our model. We will need to revise our model to ensure it will be more applicable to data outside the test set.

#### In the next cells we will separate the data into a training set and a test set using the `train_test_split()` function in scikit-learn.

When using `scikit-learn` for machine learning, we first separate the data to predictor and response variables. This is the standard way of passing datasets into a model in `scikit-learn`. The `scikit-learn` will then find out whether the predictors and responses fit the model.

In the next cell, assign the `TempAvgF` column to `y` and the remaining columns to `X`. Your `X` should be a subset of `austin_fixed` containing the following columns: 

```['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']```
 
 Your `y` should be a subset of `austin_fixed` containing one column `TempAvgF`.

In [126]:
# Your code here:
y = austin_fixed.TempAvgF
X = austin[['TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']]

In the next cell, import `train_test_split` from `sklearn.model_selection`

In [124]:
#Your code here:
from sklearn.model_selection import train_test_split as tts

Now that we have split the data to predictor and response variables and imported the `train_test_split()` function, split `X` and `y` into `X_train`, `X_test`, `y_train`, and `y_test`. 80% of the data should be in the training set and 20% in the test set. `train_test_split()` reference can be accessed [here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).


Enter your code in the cell below:

In [127]:
#Your code here:
X_train, X_test, y_train, y_test = tts(X, y, test_size= 0.2)

In [131]:
X_test

TempHighF  TempLowF  DewPointHighF  DewPointAvgF  DewPointLowF  \
819         62        47           49.0          36.0          28.0   
339         67        38           38.0          33.0          25.0   
607         84        73           74.0          71.0          66.0   
64          86        62           61.0          56.0          44.0   
263         99        77           73.0          71.0          66.0   
..         ...       ...            ...           ...           ...   
946        101        78           75.0          73.0          68.0   
665         88        62           54.0          48.0          42.0   
882         83        70           74.0          69.0          64.0   
63          77        48           57.0          46.0          30.0   
431         63        35           36.0          34.0          28.0   

     HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
819                 72.0                52.0                31.0   
339                 85.0                53.0                20.0   
607                 94.0                78.0                62.0   
64                  84.0                57.0                29.0   
263                 88.0                63.0                38.0   
..                   ...                 ...                 ...   
946                 91.0                64.0                37.0   
665                 67.0                46.0                25.0   
882                 87.0                81.0                74.0   
63                  73.0                53.0                32.0   
431                100.0                64.0                27.0   

     SeaLevelPressureHighInches  SeaLevelPressureAvgInches  ...  \
819                       30.29                      30.16  ...   
339                       30.37                      30.26  ...   
607                       30.01                      29.94  ...   
64                        30.05                      29.95  ...   
263                       29.90                      29.83  ...   
..                          ...                        ...  ...   
946                       29.99                      29.91  ...   
665                       30.30                      30.22  ...   
882                       30.07                      30.00  ...   
63                        30.05                      29.96  ...   
431                       30.03                      29.94  ...   

     VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  WindHighMPH  \
819                 10.0                10.0                10.0         18.0   
339                 10.0                10.0                10.0          9.0   
607                 10.0                 8.0                 1.0         13.0   
64                  10.0                10.0                 7.0         15.0   
263                 10.0                10.0                10.0         14.0   
..                   ...                 ...                 ...          ...   
946                 10.0                10.0                10.0         13.0   
665                 10.0                10.0                10.0          9.0   
882                 10.0                10.0                 8.0          9.0   
63                  10.0                10.0                10.0         14.0   
431                 10.0                 8.0                 3.0         13.0   

     WindAvgMPH  WindGustMPH  Snow  Fog   Rain  Thunderstorm  
819        11.0         29.0     0    0  False             0  
339         3.0         16.0     0    0  False             0  
607         4.0         22.0     0    0   True             1  
64          4.0         24.0     0    0  False             0  
263         7.0         25.0     0    0  False             0  
..          ...          ...   ...  ...    ...           ...  
946         5.0         21.0     0    0  False             0  
665         4.0         16.0     0    0  False       

#### Congratulations! Now you have finished the preparation of the dataset!

# Bonus Challenge 1

#### While the above is the common practice to prepare most datasets, when it comes to time series data, we sometimes do not want to randomly select rows from our dataset.

This is because many time series algorithms rely on observations having equal time distances between them. In such cases, we typically select the majority of rows as the test data and the last few rows as the training data. We don't use `train_test_split()` to select the train/test data because it returns random selections.

In the following cell, compute the number of rows that account for 80% of our data and round it to the next integer. Assign this number to `ts_rows`.

In [133]:
# Your code here:
ts_rows = int(len(austin_fixed)*0.8)

Assign the first `ts_rows` rows of `X` to `X_ts_train` and the remaining rows to `X_ts_test`.

In [134]:
# Your code here:
X_ts_train = X[:ts_rows]
X_ts_test = X[ts_rows:]

Assign the first `ts_rows` rows of `y` to `y_ts_train` and the remaining rows to `y_ts_test`.

In [135]:
# Your code here:

y_ts_train = y[:ts_rows]
y_ts_test = y[ts_rows:]

In [136]:
len(y_ts_test)

264